src/main/java/net/glowstone/entity/GlowLivingEntity.java

```diff
@@ -663,11 +668,60 @@ public boolean isWithinSolidBlock() {
     // Health
 
     @Override
-    public <T extends Projectile> T launchProjectile(Class<? extends T> projectile,
-        Vector velocity) {
-        T entity = world.spawn(getEyeLocation(), projectile);
-        entity.setVelocity(velocity);
-        return entity;
+    public <T extends Projectile> T launchProjectile(Class<? extends T> clazz, Vector vector) {
+        float offset = 0.0F;
+        float velocity = 1.5F;
+        T projectile = launchProjectile(clazz, vector, offset, velocity);
+        return projectile;
+    }
+
+    /**
+     * Launches a projectile from this entity.
+     *
+     * @param clazz the projectile class
+     * @param vector TODO: document this parameter
+     * @param offset TODO: document this parameter
+     * @param velocity the velocity for the first flight tick
+     * @param <T> the projectile class
+     * @return the launched projectile
+     */
+    public <T extends Projectile> T launchProjectile(Class<? extends T> clazz, Vector vector,
+            float offset, float velocity) {
+        if (vector == null) {
+            vector = getVelocity();
+        }
+
+        T projectile = throwProjectile(clazz, getEyeLocation().clone(), vector, offset, velocity);
+        projectile.setShooter(this);
+        return projectile;
+    }
+
+    protected <T extends Projectile> T throwProjectile(Class<? extends T> type, Location location,
+            Vector originalVector, float offset, float velocity) {
+        double k = Math.toRadians(-1);
+        double x = cos(k * location.getPitch()) * sin(k * location.getYaw());
+        double y = sin(k * (location.getPitch() - offset));
+        double z = cos(location.getPitch() * k) * cos(location.getYaw() * k);
+        T projectile = throwProjectile(type, location, x, y, z, velocity);
+        projectile.getVelocity().add(originalVector);
+        return projectile;
+    }
+
+    private <T extends Projectile> T throwProjectile(Class<? extends T> clazz, Location location,
+            double x, double y, double z, float velocity) {
+        double k = Math.sqrt(x * x + y * y + z * z);
```

Comment:
`k` here is the magnitude of the unit represented by `x, y, z`, previously calculated. I believe the math below multiplies the vector by the given `velocity` magnitude (using component operations) and then adds these results to the base components, in order to give a certain offset.

Reply:
Now called `magnitude` and checked for zero (so if x, y and z are all zero, we don't end up at coordinates of infinity or NaN).